src/server/schema.ts

```diff
@@ -22,37 +22,50 @@ export const schema = `
     narrower: [RelatedTerm]
     related: [RelatedTerm]
   }
-  
+
   type RelatedTerm {
     uri: ID!
     prefLabel: [String]!
   }
 
+  type Query {
+    terms(sources: [ID]!, query: String!, timeoutMs: Int = 10000): [TermsQueryResult]
+    sources: [Source]
+    lookup(uris: [ID]!, timeoutMs: Int = 10000): [LookupResult]
+  }
+
   type TermsQueryResult {
     source: Source!
     terms: [Term]! @deprecated(reason: "Use 'result' instead")
     result: TermsResult!
   }
 
-  type Query {
-    terms(sources: [ID]!, query: String!, timeoutMs: Int = 10000): [TermsQueryResult]
-    sources: [Source]
-  }
-  
   union TermsResult = Terms | TimeoutError | ServerError
-  
+
   type Terms {
     terms: [Term]
   }
-  
+
+  type LookupResult {
+    uri: ID!
+    source: Source
+    result: LookupSuccessErrorResult
```

Comment:
`result` is null when we cannot find a term for the URI. Because this isn’t an error proper, I decided to make this nullable rather than add `TermNotFound` to the union type. Does that make sense? A downside is that the lookup can be (technically) successful (i.e., a dataset was found, its distribution could be reached and there were no query errors) but still return `null` (because no term was found).

This is related to the fact that I return a `LookupResult` for _all_ IRIs in the query, even those that returned `null`. I prefer to keep this over leaving out IRIs that didn’t return anything because it’s a nice guarantee for clients that _all_ the IRIs from their query will be contained in the result (even if no terms were found).

`source` is null when we cannot find a dataset whose prefix matches the term’s IRI. Alternatively, we could combine `result` and `source` together in a new type `LookupSuccessResult`.

Reply:
Thought and read about this some more. It’s quite inelegant to have two nullable fields (both `source` and `result`). 

So how about this GraphQL schema and corresponding query structure instead?

Schema:

```graphql
type LookupQueryResult {
  uri: ID!
  source: SourceResult!
  result: LookupResult!
}

union SourceResult = Source | SourceNotFoundError

type SourceNotFoundError implements Error {
  message: String!
}

union LookupResult = Term | TimeoutError | ServerError | NotFoundError  

type TermNotFoundError implements Error {
  message: String!
}
```

Query:

```graphql
query {
  lookup(
    uris: ["https://data.rkd.nl/artists/32439", "https://data.cultureelerfgoed.nl/term/id/cht/15e29ea3-1b4b-4fb2-b970-a0c485330384"],
    timeoutMs: 1000
  ) {
    uri
    source {
      __typename
      ... on Source {
        source {
          uri
          name
          creators {
            uri
            name
            alternateName
          }
        }
      }
      ... on Error {
        message
      }
    }
    result {
      __typename
      ... on Term {
        term {
          uri
          prefLabel
          altLabel
          hiddenLabel
          scopeNote
          broader {
            uri
            prefLabel
          }
        }
      }
      ... on Error {
        message
      }
    }
  }
}`;
```

With this approach, we clearly separate the two steps in the lookup process: first find a source for the URI provided by the client, then query the source for that URI. My only gripe is that this might be too leaky an abstraction: are we bothering clients with details they aren’t interested in? Does the repeated inline fragment `... on Error { message }` overcomplicate the query?